### PR TITLE
Implementing axios-retry library

### PR DESCRIPTION
**Pull request by nikhilwaichal-96**
- Implemented the axios-retry library and created a wrapper function, axiosRetryHandler().
- Added logic to retry API calls after a certain amount of time if the error response header contains "retry-after".
- Fixed getChatToken() from retrying API calls twice due to the combined implementation of axios-retry and custom retry logic.

 proofs:
https://github.com/microsoft/omnichannel-sdk/assets/128818969/d40aa951-35d1-4836-804b-5c77ed5d7cc7

1.Retry failed api call 3 times:
image

2.Do NOT retry on 429 status code for getChatToken endpoint url
image

3. Retry on 429 for sessionInit endpoint url
image

4. Do not retry for sessionInit endpoint for ooo  (Status 400 with errorcode 705)
image

5. authcodeNonce is overwritten in error request headers
image

6. Do not retry for other endpoints with 429 status code (for example config endpoint)

image

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,8 @@ All notable changes to this project will be documented in this file.
 - Add `Ms-Oc-User-Agent` as default request headers
 - Log `RequestHeaders`
 
+- Added axios-retry library support
```
Comment (xTEddie) [inline CHANGELOG.md:21]: Changelog should be under `[Unreleased]` since we already released 0.5.1 and 0.5.0

File: src/SDK.ts [hunk 2]
```diff
@@ -229,8 +229,9 @@ export default class SDK implements ISDK {
     // construct a endpoint for anonymous chats to get LWI Details
     let requestPath = `/${OmnichannelEndpoints.LiveChatLiveWorkItemDetailsPath}/${this.omnichannelConfiguration.orgId}/${this.omnichannelConfiguration.widgetId}/${requestId}`;
     const axiosInstance = axios.create();
-    
-    axiosRetry(axiosInstance, { headerOverwrites: [OmnichannelHTTPHeaders.authCodeNonce], 
+
+    axiosRetryHandler(axiosInstance, {
+      headerOverwrites: [OmnichannelHTTPHeaders.authCodeNonce],
```
Comment (xTEddie) [inline src/SDK.ts:234]: Do we still have parity with `authCodeNonce` headers being sent for auth chats?

**Comment by xTEddie** [comment]
Is `getChatTokenRetryCount` still behaving as expected with `getChatToken()`? `getChatToken()` has its own retry mechanism as well as using `axiosRetry`. The logic was done like this to keep parity from a very old piece code. Every API calls would have up to 3 tries which is set by `maxRequestRetriesOnFailure`. However, for `getChatToken()`, it has an additional configuration with `getChatTokenRetryCount` which is set to 2 in ChatSDK which would perform additional `getChatToken` calls. So, if `maxRequestRetriesOnFailure` was set to 3 and `getChatTokenRetryCount` was set to 2, `getChatToken()` would perform up to `3*2` tries

File: src/Utils/SessionInitRetryHandler.ts [hunk 3]
```diff
@@ -1,25 +1,25 @@
...
                 }
                 break;
             case Constants.badRequestStatusCode:
-                if (parseInt(response.headers.errorcode) === Constants.outOfOfficeErrorCode) {
+                if (parseInt(error.response.headers.errorcode) === Constants.outOfOfficeErrorCode) {
                     return false;
                 }
                 break;
-            default: return true;
+            default: return isRetryableError(error) || isNetworkError(error) || error.response?.status === 408 || error.response?.status == 0 || !error.response?.status;
```
Comment (xTEddie) [inline src/Utils/SessionInitRetryHandler.ts:20]: What does `408` status code mean?

**Comment by mithunjohny** [comment]
> Is `configuration.getChatTokenRetryCount` still behaving as expected with `getChatToken()`? `getChatToken()` has its own retry mechanism as well as using `axiosRetry`. The logic was done like this to keep parity from a very old piece code. Every API calls would have up to 3 tries which is set by `maxRequestRetriesOnFailure`. However, for `getChatToken()`, it has an additional configuration with `getChatTokenRetryCount` which is set to 2 in ChatSDK which would perform additional `getChatToken` calls. So, if `maxRequestRetriesOnFailure` was set to 3 and `getChatTokenRetryCount` was set to 2, `getChatToken()` would perform up to `3*2` tries
> 
> Usually an failed API would act as follow:
> 
> 1. axios promise call
> 2. 1st call (failed)
> 3. `axiosRetry` would retry
> 4. 2nd call (failed)
> 5. `axiosRetry` would retry
> 6. 3rd call (failed)
> 7. axios promise call is rejected
> 
> As for `getChatToken()` call, it would perform as above, but it would retry with its own logic for up to 2 tries back on our configuration as soon as `axios promise call is rejected` is reached
> 
> In my opinion, the code would be more consistent and simplified if the custom logic of `getChatToken()` was removed to ensure it behaved like all the other APIs

Ok since it is expected to have additional retries we will put the code back

File: src/SDK.ts (hunk 2 above)
Comment (nikhilwaichal-96) [reply src/SDK.ts:234]: yes. we are updating headers in axiosRetryHandler function on api failure and following same logic written previously in axiosRetry custom handler function.

File: src/Utils/SessionInitRetryHandler.ts (hunk 3 above)
Comment (nikhilwaichal-96) [reply src/Utils/SessionInitRetryHandler.ts]: removed 408 status code logic

File: CHANGELOG.md (hunk 1 above)
Comment (nikhilwaichal-96) [reply CHANGELOG.md:11]: updated changelog.md file

File: src/Utils/axiosRetryHandler.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+* Custom handler for HTTP calls with Axios. Handler allows to retry HTTP calls if failed.
+*
+* @param axios Axios instance.
+* @param axiosRetryOptions Options for axios retry.
+*/
+const axiosRetryHandler = (axios: AxiosInstance, axiosRetryOptions: IAxiosRetryOptions) => { // eslint-disable-line @typescript-eslint/explicit-module-boundary-types
+
+  // Default values
+  if (axiosRetryOptions.retryOn429 === undefined || axiosRetryOptions.retryOn429 === null) {
+    axiosRetryOptions.retryOn429 = true;
```
Comment (elopezanaya) [inline src/Utils/axiosRetryHandler.ts:17]: shall we reconsiderate this and set to false by default?

File: src/Utils/SessionInitRetryHandler.ts [hunk 5]
```diff
@@ -1,25 +1,26 @@
...
+        switch (error.response.status) {
             case Constants.tooManyRequestsStatusCode:
-                if (axiosRetryOptions && axiosRetryOptions.retryOn429 === false) {
+                if (retryOn429 === false) {
                     return false;
                 }
                 break;
             case Constants.badRequestStatusCode:
-                if (parseInt(response.headers.errorcode) === Constants.outOfOfficeErrorCode) {
+                if (parseInt(error.response.headers.errorcode) === Constants.outOfOfficeErrorCode) {
```
Comment (elopezanaya) [inline src/Utils/SessionInitRetryHandler.ts:17]: shall we validate for null  the whole error object?

**Review by elopezanaya — CHANGES_REQUESTED** [review]

File: src/SDK.ts [hunk 6]
```diff
@@ -649,11 +628,11 @@ export default class SDK implements ISDK {
     const timer = Timer.TIMER();
     this.logWithLogger(LogLevel.INFO, OCSDKTelemetryEvent.SESSIONINITSTARTED, "Session Init Started", requestId);
     const axiosInstance = axios.create();
-
-    axiosRetry(axiosInstance, {
+    const retryOn429 = true;
```
[the remark this answers is not in the document]
Comment (nikhilwaichal-96) [reply src/SDK.ts:631]: for sessionInit endpoint we need to retry the failed api call so marked it as true

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@microsoft/ocsdk",
-  "version": "0.5.2-0",
+  "version": "0.5.2-0-axiosRetryTestnik",
```
[the remark this answers is not in the document]
Comment (nikhilwaichal-96) [reply package.json]: updated package.json file

File: src/SDK.ts [hunk 8]
```diff
@@ -1098,8 +1079,8 @@ export default class SDK implements ISDK {
 
     let requestPath = `/${OmnichannelEndpoints.LiveChatTranscriptEmailRequestPath}/${requestId}?channelId=${this.omnichannelConfiguration.channelId}`;
     const axiosInstance = axios.create();
-    
-    axiosRetry(axiosInstance, {
+
```
[the remark this answers is not in the document]
Comment (nikhilwaichal-96) [reply src/SDK.ts]: fixed style and removed extra lines

**Review by elopezanaya — APPROVED** [review]

**Pull request merged**